src/interface/guide/components/CastEfficiencyRibbon.tsx

```diff
@@ -0,0 +1,451 @@
+      updateType === UpdateSpellUsableType.EndCooldown;
+
+    let wastedTime = 0;
+
+    windows.forEach((window) => {
+      const windowStart = window.startTime;
+      const windowEnd = window.endTime;
+
+      const updateEvents: UpdateSpellUsableEvent[] = eventHistory.getEvents(
+        EventType.UpdateSpellUsable,
+        { spell: spell, searchBackwards: false },
+      );
+
+      updateEvents.sort((a, b) => {
+        if (a.timestamp !== b.timestamp) return a.timestamp - b.timestamp;
+        const orderA = isChargeUse(a.updateType) ? 0 : 1;
+        const orderB = isChargeUse(b.updateType) ? 0 : 1;
+        return orderA - orderB;
+      });
+
+      // Filter out simultaneous use/restore events
+      const filteredEvents: UpdateSpellUsableEvent[] = [];
+      for (let i = 0; i < updateEvents.length; i++) {
+        const event = updateEvents[i];
+        const nextEvent = updateEvents[i + 1];
+
+        if (
+          nextEvent &&
+          event.timestamp === nextEvent.timestamp &&
+          isChargeUse(event.updateType) &&
+          isChargeRestore(nextEvent.updateType)
+        ) {
+          i++;
+          continue;
+        }
+
+        filteredEvents.push(event);
+      }
+
+      // Calculate initial charges before window
+      let currentCharges = maxCharges;
+      filteredEvents.forEach((event) => {
+        if (event.timestamp < windowStart) {
+          if (isChargeUse(event.updateType)) {
+            currentCharges = Math.max(0, currentCharges - 1);
+          } else if (event.updateType === UpdateSpellUsableType.RestoreCharge) {
+            currentCharges = Math.min(maxCharges, currentCharges + 1);
+          } else if (event.updateType === UpdateSpellUsableType.EndCooldown) {
+            currentCharges = maxCharges;
+          }
+        }
+      });
+
+      const chargeEvents = filteredEvents.filter(
+        (e) => e.timestamp >= windowStart && e.timestamp <= windowEnd,
+      );
+
+      let chargesAvailable = currentCharges;
+      let segmentStart = windowStart;
+      let wasCapped = chargesAvailable === maxCharges;
+
+      chargeEvents.forEach((event) => {
+        if (wasCapped && event.timestamp > segmentStart) {
+          wastedTime += event.timestamp - segmentStart;
+        }
+
+        if (isChargeUse(event.updateType)) {
+          chargesAvailable = Math.max(0, chargesAvailable - 1);
+        } else if (event.updateType === UpdateSpellUsableType.RestoreCharge) {
+          chargesAvailable = Math.min(maxCharges, chargesAvailable + 1);
+        } else if (event.updateType === UpdateSpellUsableType.EndCooldown) {
+          chargesAvailable = maxCharges;
+        }
+
+        wasCapped = chargesAvailable === maxCharges;
+        segmentStart = event.timestamp;
+      });
+
+      if (wasCapped && windowEnd > segmentStart) {
+        wastedTime += windowEnd - segmentStart;
+      }
+    });
+
+    return wastedTime;
+  };
+
+  const wastedTime = calculateWastedTime();
```

Comment:
this needs to be memoized with `useMemo` or every re-render is going to reprocess all events for every `CastEfficiencyRibbon` rendered.

```suggestion
  const wastedTime = useMemo(calculateWastedTime(), [eventHistory, hasCharges, maxCharges]);
```

the function is pretty complicated, so you might want to move it out of the component and call it with proper parameters instead like:

```suggestion
  const wastedTime = useMemo(() => calculateWastedTime(eventHistory, ability), [ability, eventHistory]);
```